Zend/zend_string.h

```diff
@@ -330,6 +335,55 @@ static zend_always_inline void zend_string_release_ex(zend_string *s, bool persi
 	}
 }
 
+#define ZSTR_IS_LITERAL(s)	 (GC_TYPE_INFO(s) & IS_STR_LITERAL)
+#define ZSTR_IS_LITERAL_CHAR(s) (ZSTR_IS_LITERAL(s) && (ZSTR_LEN(s) <= 1))
+
+static zend_always_inline zend_string* zend_string_set_literal(zend_string *s) {
+	if (UNEXPECTED(ZSTR_IS_LITERAL(s))) {
+		return s;
+	}
+
+	if (EXPECTED(GC_REFCOUNT(s) == 1 && !ZSTR_IS_INTERNED(s))) {
+		GC_TYPE_INFO(s) |= IS_STR_LITERAL;
+		return s;
+	}
+
+	zend_string *literal = zend_string_dup(s, 0);
```

Comment:
Note that zend_string_dup doesn't duplicate interned strings.

Reply:
Woops, that was a leftover, using separate now.